pkg/graph/builder.go

```diff
@@ -714,155 +674,187 @@ func extractDependencies(env *cel.Env, expression string, resourceNames []string
+// - Template expressions (resource field values)
+// - includeWhen expressions (conditional resource creation)
+// - readyWhen expressions (resource readiness conditions)
+func validateNode(resource *Resource, templatesEnv, schemaEnv *cel.Env, resourceSchema *spec.Schema) error {
+	// Validate template expressions
+	if err := validateTemplateExpressions(templatesEnv, resource); err != nil {
+		return err
 	}
-	// include all resources, and remove individual ones
-	// during the validation
-	// this is done to avoid having to create a new context for each resource
-	for resourceName, contextResource := range resources {
-		expressionContext[resourceName] = contextResource
-	}
-
-	for _, resource := range resources {
-		// exclude resource from the context
-		delete(expressionContext, resource.id)
 
-		err := ensureResourceExpressions(env, expressionContext, resource)
-		if err != nil {
-			return fmt.Errorf("failed to ensure resource %s expressions: %w", resource.id, err)
+	// Validate includeWhen expressions if present
+	if len(resource.includeWhenExpressions) > 0 {
+		if err := validateIncludeWhenExpressions(schemaEnv, resource); err != nil {
+			return err
 		}
+	}
 
-		err = ensureReadyWhenExpressions(resource)
+	// Validate readyWhen expressions if present
+	if len(resource.readyWhenExpressions) > 0 {
+		// Create a CEL environment with only this resource's schema available
+		resourceEnv, err := krocel.TypedEnvironment(map[string]*spec.Schema{resource.id: resourceSchema})
 		if err != nil {
-			return fmt.Errorf("failed to ensure resource %s readyWhen expressions: %w", resource.id, err)
+			return fmt.Errorf("failed to create CEL environment for readyWhen validation: %w", err)
 		}
 
-		err = ensureIncludeWhenExpressions(env, includeWhenContext, resource)
-		if err != nil {
-			return fmt.Errorf("failed to ensure resource %s includeWhen expressions: %w", resource.id, err)
+		if err := validateReadyWhenExpressions(resourceEnv, resource); err != nil {
+			return err
 		}
-
-		// include the resource back to the context
-		expressionContext[resource.id] = resource
 	}
 
 	return nil
 }
 
-// ensureResourceExpressions validates the CEL expressions in the resource
-// against the resources defined in the resource graph definition.
-func ensureResourceExpressions(env *cel.Env, context map[string]*Resource, resource *Resource) error {
-	// We need to validate the CEL expressions in the resource.
-	for _, resourceVariable := range resource.variables {
-		for _, expression := range resourceVariable.Expressions {
-			_, err := ensureExpression(env, expression, []string{resource.id}, context)
+// validateTemplateExpressions validates CEL template expressions for a single resource.
+// It type-checks that expressions reference valid fields and return the expected types
+// based on the OpenAPI schemas.
+func validateTemplateExpressions(env *cel.Env, resource *Resource) error {
+	for _, templateVariable := range resource.variables {
```

Comment:
I think this could be concurrent for every single expression

Reply:
Agreed. I purposefully avoided these kind of optimizations in this PR. i want to get to logic right first, optimize later on in a seperate PR.

Side thought: curious on how much we'll win by doing concurrency here, maybe useful for very large rgds.